Autosave or update draft

oguzkocer opened the pull request:
This is a PR to discuss possible approaches to fix #10951. This issue is happening because `/autosave` endpoint has the following behavior:

* If the post is a draft, it directly updates the post. Due to a bug, while updating the post it also disables discussions on the post.
* If the post is not a draft, it creates an autosave (this is what we would have liked)

Unfortunately we are unable to address the issue in the API side, so we are left to fix it in the client in a somewhat hacky way. We'll do what the API does and update the actual draft instead of calling `/autosave`. This will fix the comments getting disabled issue and also make it obvious what's happening in the client side. (we didn't know that API was updating the actual draft until recently)

@malinajirka As discussed on Slack, I don't really know what the best way to go about this is. The approach I like the most is to encapsulate the solution as much as possible so it doesn't affect the rest of the app.

I have considered 3 types of use cases for this:
1. `AutoSavePostOrUpdateDraftUseCase`: If we were to go with this, we'd encapsulate everything in the use case and don't return a result back to `PostUploadHandler` until the whole thing is done.
2. `AutoSavePostIfNotDraftUseCase`: This is the use case I explored in this PR. The reason I tried this one over the first option is the complication around push post in `PostUploadHandler`. If we go with the first option we'll have to duplicate the push post logic or refactor that logic into a usecase and somehow chain them. Refactoring is probably a good idea, but I am not sure how complicated it'll be.
3. `FetchPostStatusUseCase`: This would be the simplest approach and we'll just return the remote post status to `PostUploadHandler`. Architecturally, this makes the most sense because it does one thing only. If we (only) do this, we can't encapsulate the complication. However, if we go with the first option, we can have this as a helper use case which would work very well.

There are other combinations of these approaches and we might even go with something completely different. I was hoping your experience with `PostUploadHandler`, however little that might be, could be helpful in deciding what makes the most sense time/value-wise.

Please keep in mind that this is a heavily WIP PR and the goal is to figure out the solution we want as fast as possible. The polish will come later and the `TODO`s in the PR points to things that I am already thinking about. That said, please feel free to comment on them, especially the question ones.

~~**_This is a discussion PR and will be closed once it serves it's purpose, so both testing steps and PR submission checklist is removed._**~~ I decided to keep this PR open since it took a lot of discussion to figure out the solution. I don't think opening a new PR would have served any real purpose after the solution was figured out.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt:
```diff
@@ -0,0 +1,108 @@
+        val localPostId = LocalId(remotePostPayload.post.id)
+        if (postStatusContinuations.containsKey(localPostId) ||
+                autoSaveContinuations.containsKey(localPostId)) {
+            // TODO: Consider canceling the previous job instead
```
Is that because we are going to use this from `PostUploadHandler`? If that's the case, I don't like a reusable component to have to rely on it's caller's implementation and I avoid it as much as I can.

malinajirka replied:
I was more thinking about throwing "IllegalStateException" with explanation that it shouldn't be invoked with the same id until the first request completes - this would work with any client. Wdyt?

oguzkocer replied:
Sure, we can do that.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt:
```diff
@@ -0,0 +1,108 @@
+            return
+        }
+        // TODO: What scope should we use for this?
+        GlobalScope.launch {
```
Yeah, I thought so too, but maybe we can inject a scope to the use case directly? Even if what ends up being injected is the `GlobalScope` I think it's worth it.

I looked at it a bit, but since it's not a priority just left a todo to explore later.

malinajirka replied:
👍 I like the idea! That would definitely make it easier to reuse and less error prone.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -604,6 +604,23 @@ private String uploadImageFile(MediaFile mediaFile, SiteModel site) {
+            finishUpload();
+        } else if (result instanceof AutoSavePostIfNotDraftResult.PostIsDraftInRemote) {
+            /**
+             * 1. If the post has a status that's not draft in local, remember that
```
Yeah, I don't know how valuable restoring the previous state is. It might be worth skipping this and give the UX a try. If the value is low, we might not do it to keep the code cleaner.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -604,6 +604,23 @@ private String uploadImageFile(MediaFile mediaFile, SiteModel site) {
+            /**
+             * 1. If the post has a status that's not draft in local, remember that
+             * 2. Set the post status to draft and push post
+             * 3. Update the local copy of post with the previous post status
```
> The local copy never gets saved into the DB, right?

I am not sure I understand this part.

> I guess we'd have to change the status, set isLocallyChanged=true and dispatch updatePost action. Wdyt?

Yeah, that's what I was thinking.

malinajirka replied:
I'd consider skipping the state restoration as you mentioned [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11251/files/4cf02b79d19bce3dd044c169d171cfad004c529a#r376649502).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -286,7 +286,7 @@ protected UploadPostTaskResult doInBackground(PostModel... posts) {
                 case REMOTE_AUTO_SAVE:
                     AppLog.d(T.POSTS, "PostUploadHandler - REMOTE_AUTO_SAVE. Post: " + mPost.getTitle());
-                    mDispatcher.dispatch(PostActionBuilder.newRemoteAutoSavePostAction(payload));
+                    mAutoSavePostIfNotDraftUseCase.autoSavePostOrUpdateDraft(payload, PostUploadHandler.this);
```
> Also what happens if the fetchPostStatus request fails? The current implementation still pushes the post, but I assume this is not what we'll want to do. Ideally we'd return what happened from the "doInBackground", right?

That's only because it's not yet implemented. I think we should figure out how we want the use case to behave first and then decide what to do with the result so we don't have to change it later.

---

For the rest of this comment, I am.. kind of confused. I re-read the comment multiple times, but I don't understand the worry about the `PUSH_POST_DISPATCHED` probably because I don't fully understand what's happening when we push post.

One very important reason for us to go with this approach would be that we can't prevent `PostUploadHandler` to handle the `OnPostUploaded` event. I mean, we can, but that'd be very complicated. So, I think we at least want push post action to be handled by `PostUploadHandler` and might even prefer to do the `autosave` there as well.

Whatever we do in the use case, it'll have a result (success or error) so as long as we handle them correctly, I don't think we can ever be in an inconsistent state.

---

@malinajirka Thank you for sharing your thoughts!

Similar to our previous Slack conversation, I think one or both of us are missing something which makes it hard to communicate effectively. Let's tackle this together on the screenshare session on Monday.

oguzkocer commented:
@malinajirka I've addressed almost everything we talked about yesterday and I think it's ready for a look. I still have documentation and unit tests to add, but to save time on those, I want to make sure we are all good with the approach and the behavior we want. One good news I have is, if we set the publish date of the post and then upload it as draft, the date is preserved and the `Schedule` button shows up in Calypso. In the app it says `Publish` but I think that's a bug. Everything seems to be working pretty nice, but I might have missed something.

I am not sure how we should test the errors, but I have added the test cases I was thinking of. Let me know what you think!

---
### Test Notes

1. Do all the tests while online
2. Then do all the tests by enabling airplane mode after opening the post list, then make the change, go back to post list and then disable the airplane mode.
3. Finally, queue multiple uploads while in airplane mode.

The results should be the same. In the first set of tests, the upload will happen as soon as the editor is closed. In the second set, it'll happen as soon as connection is available. In the third, all changes will happen as soon as connection is available one after the other.

_How to verify an autosave:_
1. Open the post in web editor which should say:
> There is an autosave of this post that is more recent than the version below. _View the autosave_
Clicking on the `View the autosave` link should show the change made.
2. Or, check stetho and make sure `/autosave` endpoint is called for that post

### Test cases:

1. Go to a remote draft's post settings
2. Change the publish date
3. Go back to post list
4. Verify that the draft is updated notification shown
5. Go to Calypso and verify that the publish date is what you have set to in Step 2 and "Schedule" button is visible (instead of "Publish")
6. Go to draft in the app and verify "Schedule" button is visible (instead of "Publish")

---

1. Open a remote draft
2. Make any change
3. Go back to post list
4. Verify that the draft is updated notification shown
5. Verify that the draft is actually updated from Calypso
6. Verify that there are no local changes for the draft in the app

---

1. Open a published post
2. Make any change
3. Go back to post list
4. Verify that `Local changes` label is visible for that post
5. Verify that the post is auto-saved

---

1. Open a privately published post (It'll be in the published list but will have a `Private` label)
2. Make any change
3. Go back to post list
4. Verify that `Local changes` label is visible for that post
5. Verify that the post is auto-saved

---

1. Open a scheduled post
2. Make any change
3. Go back to post list
4. Verify that `Local changes` label is visible for that post
5. Verify that the post is auto-saved

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt:
```diff
@@ -0,0 +1,144 @@
+        }
+    }
+
+    @Subscribe(threadMode = BACKGROUND)
```
I have made this change in bb985b123f331f84691c7101cc118ae5bc968bdf, but I couldn't test it because I can't reproduce the issue even before the change. Also, as discussed on Slack, I think there is a better way to handle the priority issue, so I'll open an issue for that once this PR is merged.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -604,6 +607,37 @@ private String uploadImageFile(MediaFile mediaFile, SiteModel site) {
+        SiteModel site = mSiteStore.getSiteByLocalId(post.getLocalSiteId());
+        if (result instanceof FetchPostStatusFailed) {
+            mPostUploadNotifier.incrementUploadedPostCountFromForegroundNotification(post);
+            mPostUploadNotifier.updateNotificationErrorForPost(post, site, post.error.message, 0);
```
Fixed in 63fd0fef5443e7b563ae2ea1e3cdc48872635898.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -604,6 +607,37 @@ private String uploadImageFile(MediaFile mediaFile, SiteModel site) {
+            finishUpload();
+        } else if (result instanceof PostAutoSaveFailed) {
+            mPostUploadNotifier.incrementUploadedPostCountFromForegroundNotification(post);
+            mPostUploadNotifier.updateNotificationErrorForPost(post, site, post.error.message, 0);
```
Fixed in 63fd0fef5443e7b563ae2ea1e3cdc48872635898.

oguzkocer commented:
@malinajirka Thank you for the review. I've fixed the issues you mentioned and replied to your comments with the commit hashes. I've then added documentation in 1268a78cab7f67c5dfa8f11ecc3d79b2fab0ba30. Finally, I worked on the unit tests in all the remaining commits (between 174b742b0d40226a6f848d837542ec035375b9f8 and 26db1f9e3b6ea4882c73f4a65124dd6d1d5fdcd7) which are contained in one file`AutoSavePostIfNotDraftUseCaseTest`, so you can review that separately. The only case I didn't intentionally cover in the unit tests is the `IllegalArgumentException` when a post is already being processed as I don't think it'll provide enough value.

Let me know what you think!

oguzkocer commented:
Thanks @malinajirka! I intended to close this PR and open a fresh one when we figured out the approach. However, that happened fairly late in the PR, so I think opening a fresh one would just confuse us if we need to refer back to it.
